src/Esprima/Ast/List.cs

```diff
@@ -0,0 +1,153 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Esprima.Ast
+{
+    // This structure is like List<> from the BCL except it is designed
+    // to be modifiable by this library during the construction of the AST
+    // but publicly read-only thereafter. The only allocation required on the
+    // heap is the backing array storage for the element. An empty list,
+    // however causes no heap allocation; that is, the array is allocated
+    // on first addition.
+
+    public struct List<T> : IReadOnlyList<T>
```

Comment:
Maybe this should be called NodeList/AstList/similar so there's clear distinction between BCL one and this?

Reply:
I completely agree. In fact, I completely forgot to say in the opening comments that I'm not happy with the name but it's odd to call it a `NodeList` when `T` is unconstrained.

Reply:
I'm afraid that won't work as the model stands right now, unless the relations are not all completely thought out. Just to cite one example, `BlockStatement.Body` is a list of `StatementListItem` objects where the `StatementListItem` _interface_ has no relation to `INode`. There many other such instances.